src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala

```diff
@@ -42,26 +51,38 @@ object InstanceUpdater extends StrictLogging {
   private def shouldBeExpunged(instance: Instance): Boolean =
     instance.tasksMap.values.forall(t => t.isTerminal) && instance.state.goal == Goal.Decommissioned
 
+  private def shouldAbandonReservation(instance: Instance): Boolean = {
+
+    def allAreGoneByOperator = instance.tasksMap.values.iterator
+      .flatMap(_.status.mesosStatus)
+      .forall(_.getState == MesosProtos.TaskState.TASK_GONE_BY_OPERATOR)
```

Comment:
For pods, this is be problematic. If you have a pod with 2 tasks, one can end up TASK_FINISHED **without tearing down the whole pod**. I.e., the other task will just keep on running or whatever.

So when the agent drops two days later, and the operator marks the agent as gone, you'd have (TASK_FINISHED, TASK_GONE_BY_OPERATOR). 

Therefore, I suggest to abandon the reservation if at least one task is TASK_GONE_BY_OPERATOR (`.exists` instead of `forall`) :)

We might want to confirm though that all tasks are either terminal or TASK_GONE_BY_OPERATOR – or to be more conservative, either TASK_FINISHED or TASK_GONE_BY_OPERATOR.

Reply:
@meichstedt yeah, this is very related to the problem of using TASK_GONE_BY_OPERATOR in the first place.

It was `.exists` at first but then there is a possibility where we rewind abandon a reservation multiple times, and this would cause us to abandon a newly created reservation.